File: 347. Top K Frequent Elements/Java/Solution.java
public class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freqDict = new HashMap<>();
        int maxFreq = -1;
        for (int num : nums) {
            if (!freqDict.containsKey(num)) {
                freqDict.put(num, 0);
            }
            int freq = freqDict.get(num) + 1;
            freqDict.put(num, freq);
            maxFreq = Math.max(maxFreq, freq);
        }
        List<Integer>[] freqList = new List[maxFreq + 1];
        for (Map.Entry<Integer, Integer> entry : freqDict.entrySet()) {
            if (freqList[entry.getValue()] == null) {
                freqList[entry.getValue()] = new ArrayList<>();
            }
            freqList[entry.getValue()].add(entry.getKey());
        }
        List<Integer> result = new ArrayList<>();
        for (int i = freqList.length - 1; i > 0 && result.size() < k; i--) {
            if (freqList[i] != null) {
                result.addAll(freqList[i]);
            }
        }
        return result;
    }
}

File: 347. Top K Frequent Elements/C++/Solution.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int> &nums, int k) {
        unordered_map<int, int> freqDict;
        int maxFreq = -1;
        for (auto num : nums) {
            if (freqDict.find(num) == freqDict.end()) {
                freqDict[num] = 0;
            }
            freqDict[num]++;
            maxFreq = max(maxFreq, freqDict[num]);
        }
        vector<vector<int>> freqList(maxFreq + 1);
        for (auto item : freqDict) {
            freqList[item.second].push_back(item.first);
        }
        vector<int> result;
        for (int i = freqList.size() - 1; i > 0 && result.size() < k; i--) {
            if (!freqList[i].empty()) {
                result.insert(result.end(), freqList[i].begin(), freqList[i].end());
            }
        }
        return result;
    }
};